### PR TITLE
[Relay][Pass] Add pass to remove unused functions in relay module

### PR DESCRIPTION
**Pull request by wweic**
https://discuss.tvm.ai/t/remove-unused-function-in-relay-module/4753

cc @icemelon9 @zhiics @jroesch @tqchen @MarisaKirisame

### CODE REVIEW

**Comment by MarisaKirisame** [comment]
can you make the function accept a set of root function? they will not get removed under any circumstance.
by default it is the main function.

**Comment by weberlo** [comment]
@MarisaKirisame is [dead code elimination](https://github.com/apache/incubator-tvm/blob/master/src/relay/pass/dead_code.cc) still broken? If not, we should try to merge these two passes, and add a few configuration knobs (e.g., `interprocedural`, `intraprocedural`).

**Comment by wweic** [comment]
@MarisaKirisame Changed as you suggested. Thank you!

**Comment by zhiics** [comment]
Yes, we do need have a more powerful DCE pass. I think we may need to have some liveness analysis late on since @jroesch will probably need them for memory planning.

**Comment by MarisaKirisame** [comment]
Dead Code Elimination is still broken until higher order pointer analysis is there. I will get back to it after the pldi dead line.

File: src/relay/backend/vm/compiler.cc [hunk 1]
```diff
@@ -863,6 +864,8 @@ void VMCompiler::Compile(Module mod,
 
 Module VMCompiler::OptimizeModule(const Module& mod, const TargetsMap& targets) {
   Array<Pass> pass_seqs;
+  Array<tvm::Expr> entry_functions{tvm::Expr{"main"}};
+  pass_seqs.push_back(transform::RemoveUnusedFunctions(entry_functions));
```
Comment (icemelon9) [inline src/relay/backend/vm/compiler.cc]: I wonder if we should run this pass at the end of optimization. For example, some lambda lifted functions will be no longer used after they are inlined.

Comment (wweic) [reply src/relay/backend/vm/compiler.cc]: This is a good idea, we can run it twice begin and end. But I check `InlinePrimitives`, it doesn't seem to inline identity function in test `test_remove_all_prelude_functions_but_referenced_functions`. I can fix it later.

**Comment by zhiics** [comment]
Thanks @wweic @jroesch @MarisaKirisame

**Pull request merged**